File: src/lib/studio/engine/buses.ts
// Runtime wrapper for one session Bus — mirror of EngineTrack, minus
// clip scheduling. A bus is a stereo submix: things (tracks or other
// buses) connect INTO its `input` node; the bus fader + pan + inserts
// process the mixed signal; the tail (`output`) connects to whatever
// downstream target the session declares (another bus or master).
//
// Solo semantics extend the track logic: while ANY bus or track is
// soloed, non-soloed buses whose downstream tail doesn't feed a
// soloed source are muted. The engine computes this globally.

import * as Tone from 'tone';
import type { Bus } from '../session';
import { buildFxChain, type EngineFxChain } from './fx';

export interface EngineBus {
  busId: string;
  /** Where other tracks/buses connect INTO this bus. */
  input: Tone.ToneAudioNode;
  /** The tail of the strip — connect this to a downstream target
   *  (another bus's input, or the engine's masterIn). */
  output: Tone.ToneAudioNode;
  /** Strip's PanVol node — exposed so automation (Phase 8) can drive
   *  .volume / .pan directly. Not for signal-graph reconnection. */
  panvol: Tone.PanVol;
  userMute: boolean;
  userSolo: boolean;
  updateStrip: (patch: { volume_db?: number; pan?: number; mute?: boolean; solo?: boolean }) => void;
  dispose: () => void;
}

export function buildBus(bus: Bus): EngineBus {
  const panvol = new Tone.PanVol(bus.pan, bus.volume_db);
  const muteGate = new Tone.Gain(bus.mute ? 0 : 1);
  const fx = buildFxChain(bus.fx);

  panvol.connect(muteGate);
  muteGate.connect(fx.input);

  const disposers: Array<() => void> = [
    () => panvol.dispose(),
    () => muteGate.dispose(),
    () => fx.dispose(),
  ];

  return {
    busId: bus.id,
    input: panvol,
    output: fx.output,
    panvol,
    userMute: bus.mute,
    userSolo: bus.solo,
    updateStrip: (patch) => {
      const now = Tone.now();
      if (patch.volume_db !== undefined) {
        try { panvol.volume.cancelScheduledValues(now); panvol.volume.linearRampTo(patch.volume_db, 0.03, now); }
        catch { panvol.volume.value = patch.volume_db; }
      }
      if (patch.pan !== undefined) {
        try { panvol.pan.cancelScheduledValues(now); panvol.pan.linearRampTo(patch.pan, 0.03, now); }
        catch { panvol.pan.value = patch.pan; }
      }
      if (patch.mute !== undefined) {
        try { muteGate.gain.cancelScheduledValues(now); muteGate.gain.linearRampTo(patch.mute ? 0 : 1, 0.02, now); }
        catch { muteGate.gain.value = patch.mute ? 0 : 1; }
      }
      // solo handled by caller (engine-level)
    },
    dispose: () => {
      for (const d of disposers) try { d(); } catch { /* ignore */ }
    },
  };
}

// Silence the unused type-import warning on downstream files.
export type _EngineFxChainMarker = EngineFxChain;

File: ios/App/App/Studio/Buses.swift
// Buses — Swift mirror of src/lib/studio/engine/buses.ts.
//
// One user bus at runtime: a stereo submix that tracks (or other
// buses) route INTO via `strip`, and whose `output` node feeds the
// downstream target the session declares (another bus's strip, or
// the engine's masterMixer).
//
// Same graph mutation rules as TrackBinding: connections here are
// only safe while the AVAudioEngine is STOPPED. Engine.loadSession
// wraps the whole rebuild with a stop / restart pair.

import AVFoundation

public final class BusBinding {
    public let busId: String
    public var userMute = false
    public var userSolo = false

    private let engine: AVAudioEngine
    /// Input into this bus. Tracks / other buses connect their tail
    /// to `strip` when routed here.
    public let strip: AVAudioMixerNode      // pan + vol
    private let muteGate: AVAudioMixerNode  // 0/1 gain
    private var fxChain: FxChain?
    /// The last node in the bus's internal chain. Downstream targets
    /// (another bus's strip, or the engine's masterMixer) connect
    /// from THIS node. Engine.loadSession wires that link once every
    /// bus is built.
    public var output: AVAudioNode

    private init(busId: String,
                 engine: AVAudioEngine,
                 strip: AVAudioMixerNode,
                 muteGate: AVAudioMixerNode,
                 fxChain: FxChain?,
                 output: AVAudioNode)
    {
        self.busId = busId
        self.engine = engine
        self.strip = strip
        self.muteGate = muteGate
        self.fxChain = fxChain
        self.output = output
    }

    /// Attach + wire the bus's internal chain: strip → muteGate →
    /// [fx chain] → output. `output` is exposed on the returned
    /// binding so Engine.loadSession can wire it into whatever the
    /// bus's declared downstream target is (bus vs master).
    public static func build(bus: Studio.Bus, engine: AVAudioEngine) -> BusBinding {
        let strip = AVAudioMixerNode()
        let muteGate = AVAudioMixerNode()
        engine.attach(strip)
        engine.attach(muteGate)

        strip.outputVolume = Float(dbToGain(bus.volume_db))
        strip.pan = Float(bus.pan)
        muteGate.outputVolume = bus.mute ? 0 : 1

        engine.connect(strip, to: muteGate, format: nil)
        let fxChain = FxChain.build(engine: engine, specs: bus.fx)
        let tail: AVAudioNode
        if let chain = fxChain {
            engine.connect(muteGate, to: chain.input, format: nil)
            tail = chain.output
        } else {
            tail = muteGate
        }

        let binding = BusBinding(busId: bus.id, engine: engine,
                                 strip: strip, muteGate: muteGate,
                                 fxChain: fxChain, output: tail)
        binding.userMute = bus.mute
        binding.userSolo = bus.solo
        return binding
    }

    // MARK: - Live strip control (Phase 6-equivalent — bus strip
    // fader/pan/mute apply without a full engine rebuild).

    public func setVolumeDb(_ db: Double) { strip.outputVolume = Float(dbToGain(db)) }
    public func setPan(_ p: Float) { strip.pan = max(-1, min(1, p)) }
    public func setMute(_ m: Bool) { muteGate.outputVolume = m ? 0 : 1 }

    public func dispose() {
        // Order matches TrackBinding.dispose: detach nodes first,
        // then release the fx chain so nothing else references it.
        engine.detach(strip)
        engine.detach(muteGate)
        fxChain?.dispose()
        fxChain = nil
    }
}
